habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomMonorepoGroupMigration.java

```diff
@@ -0,0 +1,147 @@
+            if (toolPoetryDependencies.isPresent()) {
+                Config foundDependencies = toolPoetryDependencies.get();
+                Map<String, Object> dependencyMap = foundDependencies.valueMap();
+
+                for (Map.Entry<String, Object> dependency : dependencyMap.entrySet()) {
+                    String packageName = dependency.getKey();
+                    Object packageRhs = dependency.getValue();
+                    if (TomlUtils.representsLocalDevelopmentVersion(packageRhs)) {
+                        String packageRshAsString = TomlUtils.convertCommentedConfigToToml((CommentedConfig) packageRhs);
+                        logger.info("Found local dependency not within monorepo group! ({} = {})", packageName, packageRshAsString);
+                        TomlReplacementTuple replacementTuple = new TomlReplacementTuple(packageName, packageRshAsString, "");
+                        replacements.put(packageName, replacementTuple);
+                        shouldExecute = true;
+                    }
+                }
+            }
+
+            Optional<Config> toolPoetryMonorepoDependencies = tomlFileConfig.getOptional(TomlUtils.TOOL_POETRY_GROUP_MONOREPO_DEPENDENCIES);
+            if (toolPoetryMonorepoDependencies.isPresent()) {
+                hasExistingMonoRepoDependenciesGroup = true;
+            }
+        }
+
+        return shouldExecute;
+    }
+
+    @Override
+    protected boolean performMigration(File pyProjectTomlFile) {
+        String fileContent = StringUtils.EMPTY;
+        try (BufferedReader reader = new BufferedReader(new FileReader(pyProjectTomlFile))) {
+            String line = reader.readLine();
+            boolean injectAfterNextEmptyLine = false;
+
+            while (line != null) {
+                boolean addLine = true;
+                boolean isEmptyLine = line.isBlank();
+
+                if (line.contains(StringUtils.SPACE) || line.contains(TomlUtils.EQUALS)) {
+                    String key = line.substring(0, line.indexOf(StringUtils.SPACE));
+
+                    if (key == null) {
+                        key = line.substring(0, line.indexOf(TomlUtils.EQUALS));
+                    }
+
+                    if (key != null) {
+                        key = key.strip();
+
+                        TomlReplacementTuple matchedTuple = replacements.get(key);
+                        if (matchedTuple != null) {
+                            // skip this line, we will add it back to [tool.poetry.group.monorepo.dependencies] later
+                            addLine = false;
+                        }
+                    }
+
+                } else if (line.contains("[") && line.contains("]")) {
+                    String key = line.strip();
+
+                    if (hasExistingMonoRepoDependenciesGroup && (key.equals("[" + TomlUtils.TOOL_POETRY_GROUP_MONOREPO_DEPENDENCIES + "]"))) {
```

Comment:
A: We use this same string multiple times (including the very next condition).  For readability, maybe extract to a var or a string-builder method?

Reply:
The next condition is actually a different group id. (monorepo dependencies vs. normal dependencies)

This is one of these scenarios where nothing is super readable.  Could add a method that only ever gets executed once, but i tend to think what is there is more readable in the situation (though certainly arguable).  I borrowed this from another location in Habushu and remember having this same debate with I PR-ed that code - but ultimately was okay with it.